pkgs/stdenv/generic/check-meta.nix

```diff
@@ -248,45 +244,67 @@ let
       isEnabled = lib.findFirst (x: x == reason) null showWarnings;
     in if isEnabled != null then builtins.trace msg true else true;
 
-  metaTypes = with lib.types; rec {
+
+  # Some surface-level type check. We are using NixOS'
+  # option types here, which however have the major drawback
+  # of not providing full type checking (part of the type check is
+  # done by the module evaluation itself). Therefore, the checks
+  # will not recurse into attributes.
+  # We still provide the full type as a comment for documentation
+  # purposes and in the hope that they will be used eventually.
+  # See https://github.com/NixOS/nixpkgs/pull/191171 for an attempt
+  # to fix this, or mkOptionType in lib/types.nix for more information.
+  metaTypes = with lib.types; let
+    /* In `types.nix` there is a `attrs` type but no `list` type. Since those
+     * are too generic and should not be used under normal circumstances, declare
+     * the type locally in here.
+     */
```

Comment:
The only difference is that in the error message `listOf anything` would show instead of `list`, which I thought misleading. So I could equally do `list = (listOf anything) // { description = "list"; }` if you prefer to.

Reply:
I think the error message should show the actual type, e.g. `list of string`, even if we don't check the "of string" part. In other words just keep the types as they were.